### PR TITLE
client: limit spider scope and use own proxies

### PR DESCRIPTION
**Pull request by thc202**
Limit to context, subtree, or target preventing any accesses outside the selected spider scope.
Use own proxies when spidering instead of the main one, to limit the scope and allow later to track the requests of each action.

### CODE REVIEW

**Comment by psiinon** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [340aab5f\-5503\-4756\-b567\-f104510599a7](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=340aab5f-5503-4756-b567-f104510599a7&branch=thc202%2Fclient%2Fspider%2Flimit-scope)
### New Issues
Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---
LOW|[Log_Forging](https://eu.ast.checkmarx.net/results/0da811fa-6373-42b1-bbb6-d0026285b136/340aab5f-5503-4756-b567-f104510599a7/sast/description/117/9884244256835554923)|/addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderDialog\.java: [342](https://github.com/thc202/zap-extensions/blob/client/spider/limit-scope//addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderDialog.java#L342)|[Attack Vector](https://eu.ast.checkmarx.net/results/340aab5f-5503-4756-b567-f104510599a7/0da811fa-6373-42b1-bbb6-d0026285b136/sast?result-id=BGHeoWHZOHZAb2xUutWUfjfzv1o%3D&redirect=true)

**Review by psiinon — APPROVED** [review]

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java [hunk 1]
```diff
@@ -577,7 +676,194 @@ public TableModel getActionsTableModel() {
...
+                    .append(HttpHeader.CONTENT_LENGTH)
+                    .append(": ")
+                    .append(outOfScopeResponseBody.length())
+                    .append(crlf);
+
+            HttpResponseHeader responseHeader;
+            try {
+                responseHeader = new HttpResponseHeader(strBuilder.toString());
+            } catch (HttpMalformedHeaderException e) {
+                LOGGER.error("Failed to create a valid! response header: ", e);
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java:785]: Drop the exclamation point?

Comment (thc202) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java]: Done.

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/MessagesTable.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    private static class ProcessedCellItemIconHighlighter extends AbstractHighlighter {
+
+        private static final ImageIcon ALLOWED_ICON =
+                DisplayUtils.getScaledIcon(
+                        MessagesTable.class.getResource("/resource/icon/16/152.png"));
+
+        private static final ImageIcon NOT_ALLOWED_ICON =
+                DisplayUtils.getScaledIcon(
+                        MessagesTable.class.getResource("/resource/icon/16/149.png"));
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/spider/MessagesTable.java:113]: Could we give the icons useful filenames?

I know this is a holdover from code/history but I think it'll be more maintenance friendly in the long run.

Comment (kingthorin) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/MessagesTable.java:113]: I guess these already exist, so maybe that's a change for another PR.

**Review by kingthorin — APPROVED** [review]

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/MessagesTable.java (hunk 2 above)
Comment (thc202) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/MessagesTable.java:113]: Those are core icons.

**Pull request merged**